src/Middleware/Rewrite/src/RewriteBuilderExtensions.cs

```diff
@@ -46,7 +48,45 @@ public static IApplicationBuilder UseRewriter(this IApplicationBuilder app, Rewr
             }
 
             // put middleware in pipeline
-            return app.UseMiddleware<RewriteMiddleware>(Options.Create(options));
+            return AddRewriteMiddleware(app, Options.Create(options));
+        }
+
+        private static IApplicationBuilder AddRewriteMiddleware(IApplicationBuilder app, IOptions<RewriteOptions>? options)
+        {
+            const string globalRouteBuilderKey = "__GlobalEndpointRouteBuilder";
+            // Check if UseRouting() has been called so we know if it's safe to call UseRouting()
+            // otherwise we might call UseRouting() when AddRouting() hasn't been called which would fail
+            if (app.Properties.TryGetValue("__EndpointRouteBuilder", out _) || app.Properties.TryGetValue(globalRouteBuilderKey, out _))
+            {
+                return app.Use(next =>
+                {
+                    if (options is null)
+                    {
+                        options = app.ApplicationServices.GetRequiredService<IOptions<RewriteOptions>>();
+                    }
+
+                    app.Properties.TryGetValue(globalRouteBuilderKey, out var routeBuilder);
+                    // start a new middleware pipeline
+                    var builder = app.New();
+                    builder.UseMiddleware<RewriteMiddleware>(options);
+                    if (routeBuilder is not null)
+                    {
+                        // use the old routing pipeline if it exists so we preserve all the routes and matching logic
+                        builder.Properties[globalRouteBuilderKey] = routeBuilder;
+                    }
+                    builder.UseRouting();
+                    builder.Run(next);
```

Comment:
Same comment as above, we need to pass this branched pipeline to the RewriteMiddleware so that we don't always run this branch. It's only run when there's a path rewrite.